test/test.ts

```diff
@@ -73,96 +84,275 @@ describe('Retrier', () => {
+      backoff.backoff();
+      backoff.reset();
 
+      mockClock.tick(10);
+      expect(readyCallback).not.to.have.been.called;
+    });
+
+    it('backoff should be reset after fail', () => {
+      sinon.spy(backoff, 'reset');
+
+      backoff.failAfter(1);
+      backoff.backoff();
+      mockClock.tick(10);
+      backoff.backoff();
+
+      expect(backoff.reset).to.have.been.calledOnce;
+      backoff.reset.restore();
+    });
+
+    it('backoff limit should be greater than 0', () => {
+      try {
+        backoff.failAfter(0);
+        expect.fail('Calling backoff.failAfter with 0 should have thrown an error!');
+      } catch (err) {
+        expect(err.message).to.equal('Expected a maximum number of retry greater than 0 but got 0');
+      }
+    });
+
+    it('the backoff number should increase from 0 to N - 1', () => {
+      const expectedNumbers = [0, 1, 2, 3, 4];
+
+      for (let i of expectedNumbers) {
+        backoff.backoff();
+        mockClock.tick(backoff.next());
+        expect(backoffCallback.getCall(i).args[0]).to.equal(i);
+      }
+    });
+
+    it('the randomisation factor should be between 0 and 1', () => {
+      testFunc({randomisationFactor: 1.1}, 'The randomisation factor must be between 0 and 1.');
+      testFunc({randomisationFactor: -0.1}, 'The randomisation factor must be between 0 and 1.');
+      testFunc({randomisationFactor: 0}, 'The randomisation factor must be between 0 and 1.');
+      testFunc({randomisationFactor: 1}, 'The randomisation factor must be between 0 and 1.');
+    });
+
+    it('the raw delay should be randomized based on the randomisation factor', () => {
+      backoff = Backoff.exponential({
+        randomisationFactor: 0.5,
+        initialDelay: 10,
+        maxDelay: 600
+      });
+      for (let i = 0; i < 10; i++) {
+        let next = backoff.next();
+        expect(backoff.next()).to.be.below(601);
+        expect(backoff.next()).to.be.at.least(10);
+      }
+    });
+
+    it('the initial backoff delay should be greater than 0', () => {
+      testFunc({initialDelay: -0.1}, 'The initial timeout must be greater than 0.');
+      testFunc({initialDelay: 0}, 'The initial timeout must be greater than 0.');
+    });
+
+    it('the maximal backoff delay should be greater than 0', () => {
+      testFunc({maxDelay: -0.1}, 'The maximal timeout must be greater than 0.');
+      testFunc({maxDelay: 0}, 'The maximal timeout must be greater than 0.');
+    });
+
+    it('the maximal backoff delay should be greater than the initial backoff delay', () => {
+      testFunc({
+        maxDelay: 5,
+        initialDelay: 10
+      }, 'The maximal backoff delay must be greater than the initial backoff delay.');
+
+    });
+
+    it('delays should follow an exponential sequence', () => {
+      const expectedDelays = [10, 20, 40, 80, 160, 320, 640, 1000, 1000];
+      for (let delay of expectedDelays) {
+        expect(backoff.next()).to.equal(delay);
```

Comment:
Currently we have both randomization broken AND backoff in tests is initialized with empty randFactor which makes it 0 by default. Pls make sure that initializing backoff with non-zero randFactor will fail this test (manually, to make sure randomization works).